Fixes #32166 - reworded ignored interface setting

domitea opened the pull request:
Foreman has two lists where "ignored" facts can be defined. The Ignored Interface Identifiers and Excluded Facts. Both of them actually do almost the same thing so it makes sense to merge them. 

PR contains also migration that takes values from ignored interfaces to excluded facts.
I've in plan to check plugins if the ignored interfaces are used and create PRs to reflect these change. In addition to that. I've added a warning message for Setting call and for Template call of this settings just in case if there are could be something missing from community. 


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

ekohl commented on app/models/setting/provisioning.rb:
```diff
@@ -59,7 +59,6 @@ def self.default_settings
       set('access_unattended_without_build', N_("Allow access to unattended URLs without build mode being used"), false, N_('Access unattended without build')),
       set('manage_puppetca', N_("Foreman will automate certificate signing upon provision of new host"), true, N_('Manage PuppetCA')),
       set('ignore_puppet_facts_for_provisioning', N_("Stop updating IP address and MAC values from Puppet facts (affects all interfaces)"), false, N_('Ignore Puppet facts for provisioning')),
-      set('ignored_interface_identifiers', N_("Ignore interfaces that match these values during facts importing, you can use * wildcard to match names with indexes e.g. macvtap*"), IGNORED_INTERFACES, N_('Ignore interfaces with matching identifier')),
```
The constant `IGNORED_INTERFACES` is no unused. Should it be removed?

ekohl commented on config/environments/development.rb:
```diff
@@ -46,7 +46,7 @@
   if defined?(Bullet)
     config.after_initialize do
-      Bullet.enable = true
+      Bullet.enable = false
```
This looks unrelated

ekohl reviewed (changes requested):
I'm not sure this is a good idea. In Facter there's structured facts unless the fact filter starts to understand that structure as well.

Back in Facter < 3 you have facts as `ipaddress_$interface` but in Facter 3, there's a hash, so you end up with `facts['networking']['interfaces'][$interface]`. Having a list of interface patterns to ignore is actually very useful. A fact parser can use this information to filter out the right facts.

There is actually also a bug with this global matching where some facts are matched by accident. For example, `en*v*` also filters out the `environment` fact since it matches. I'm not sure if there's a Redmine for that though.

domitea commented:
I understand your concern but I think it makes sense to have only one list of fact filtering. Information about interfaces are facts too so this attempt to remove inconsistency. To be honest, when I viewed the problem for the first time I through there is a another "magic parser and filter" only for the interfaces. Then @ares comes and said that it's only a another list of filters for fact parser. 

From my point of view, the `en*v*` problem isn't problem at all because Regex. Maybe, but maybe it will worth to create another list of "must have" facts that will bypass the filter even the regex in excluded facts includes them because sometimes is simpler to define what do you want instead you don't want. 

What so you think?

lzap commented:
I also do not think this is a good idea, but for a different reason. A do believe that the best user experience will bring us Common Fact Model: https://community.theforeman.org/t/rfc-common-fact-model/18900

In short, instead of filtering out facts, we want to have highly curated list of facts that we actually accept. We are speaking about dozen of facts at maximum, but these would be normalized across all facters (Ansible, Puppet, RHSM) in a way they do not fight each other. In some cases we could actually drop fact type completely and only present one consistent fact (e.g. memory size). All facts would be also stored in JSON as blobs so they could be still used per-host for parametrization (each type separately).

Then we can drop fact filtering altogether, there is no reason to keep it because storing a JSON in a blob is N operation. And here is my point - if we merge fact filter and NIC filter, then we could not drop the former because NIC filtering will still be useful. Our fact parsers will still be parsing NICs even after we introduce Common Fact Model.

So it is not a good step for the future. I am assuming that people do like my proposal, I know, it's bold :-)

ekohl commented:
> I understand your concern but I think it makes sense to have only one list of fact filtering. Information about interfaces are facts too so this attempt to remove inconsistency. To be honest, when I viewed the problem for the first time I through there is a another "magic parser and filter" only for the interfaces. Then @ares comes and said that it's only a another list of filters for fact parser.

Interfaces are special anyway since there is code in Foreman that updates the host interfaces from facts. That alone makes them special enough that a specific filter is valid IMHO.

At this point I don't think this is a path we should take and perhaps never should as long as the interface updating code exists in Foreman.

ares commented:
> Interfaces are special anyway since there is code in Foreman that updates the host interfaces from facts. That alone makes them special enough that a specific filter is valid IMHO.
> 
> At this point I don't think this is a path we should take and perhaps never should as long as the interface updating code exists in Foreman.

I think you're right. I think we can avoid some user confusion by providing a better help to the existing settings instead of combining them.

The `excluded_facts` means it skips all facts from being imported, meaning stored as raw data in the Foreman DB. The existing help for that says 

```
Exclude pattern for all types of imported facts (puppet, ansible, rhsm). Those facts won't be stored
in foreman's database. You can use * wildcard to match names with indexes e.g. ignore* will filter
out ignore, ignore123 as well as a::ignore or even a::ignore123::b
```

Which I think is good enough.

The `ignore_interace_identifiers` one is used after the facts are stored in the DB and we do further processing and update existing Foreman objects accordingly. The help says today 

```
Ignore interfaces that match these values during facts importing, you can use * wildcard 
to match names with indexes e.g. macvtap*
```

I think we should update it to something like

```
Skip creating or updating host network interfaces objects with identifiers matching these values from incoming facts.
You can use * wildcard to match identifiers with indexes e.g. macvtap*. The ignored interfaces raw facts
will be still stored in the DB, see the 'Exclude pattern' setting for more details.
```

If you all agree, I think this PR can just change to modify the setting description. As of today that would require a migration too, which hopefully changes soon.

lzap commented:
> If you all agree, I think this PR can just change to modify the setting description. As of today that would require a migration too, which hopefully changes soon.

I agree, let's do this. I understand the concern that this is confusing. Ideally at some point in the future:

1) Facts are stored in a blob/text and only few stored in tables separately
2) Foreman does not create interfaces, subnets, domains automatically but only via human-invervention

This is what I believe would deliver the best experience.

domitea committed:
Fixes #32166 - Merging Ignored Interfaces to Excluded Facts

Foreman has two lists where "ignored" facts can be defined.
The Ignored Interface Identifiers and Excluded Facts. Both of them
actually do almost the same thing so it makes sense to merge them.

domitea commented:
@lzap, @ares PR updated, please take a look.

lzap reviewed (approved):
Works fine, thanks.

lzap commented:
@ekohl please merge once you confirm.

ekohl reviewed (approved):
LGTM. @ares since you also took a look, want to take a last one before merging?

ares reviewed (approved):
Thanks, merging!

ares commented:
I've updated the commit message as it was no longer accurate.

lzap commented:
We have missed the commit message which was not describing the actual commit, renamed the PR at least.

ezr-ondrej commented on db/migrate/20210610134000_update_ignored_interfaces_setting_description.rb:
```diff
@@ -0,0 +1,8 @@
+class UpdateIgnoredInterfacesSettingDescription < ActiveRecord::Migration[6.0]
+  def up
+    setting = Setting.find_by(name: 'ignored_interface_identifiers')
+    return unless setting
+
+    setting.update(description: N_("Skip creating or updating host network interfaces objects with identifiers matching these values from incoming facts. You can use * wildcard to match identifiers with indexes e.g. macvtap*. The ignored interfaces raw facts will be still stored in the DB, see the 'Exclude pattern' setting for more details."))
+  end
+end
```
We do not need migrations for description. I'll clean it up, but unfortunatelly it's already in release :)
